### PR TITLE
rds: ensure fulfillment of requested rds resources

### LINKED ISSUE #3258: XDS state machine results in continuous looping when updated listeners are yet to drain
**Bug description**:
With recent changes to the XDS state machine `pkg/envoy/ads/stream.go`, there is a scenario where the XDS state machine can enter in a continuous loop of requesting RDS resources referenced in a stale listener that has not yet been drained. The default drain interval is 600s (10 minutes), which means when a listener is updated to remove previously set RDS references, Envoy will continuously request for this resource till the older listener configuration is completely drained.

Such a scenario arises for ingress (and egress in the future) configurations where the filter chains referencing route configurations can be added and removed dynamically based on policies applied to the cluster. When a policy is removed and the corresponding listener is updated, Envoy continuously requests for stale RDS resources which no longer can be generated because of the absence of corresponding policies.

This issue proposes to update the default drain interval from 600s to 5s, so listeners are drained as soon as possible upon reconfiguration so that Envoy stops requesting stale resources while the listeners are being drained. Additionally, a default deny filter_chain could be necessary on the listener to account for cases where there are no filter chains on the listener (no service filter chains + no egress filter chain + no ingress filter chain), otherwise, the updated LDS response will exclude the listener which will require supporting unsubscriptions for resources in the XDS state machine.

**Affected area (please mark with X where applicable)**:
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [ ]
- Egress                     [ ]
- Envoy Control Plane        [X]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ ]
- Tests                      [ ]
- Demo                       [ ]
- CI System                  [ ]

### PR DESCRIPTION
**Pull request by eduser25**
Changes in LDS could leave listeners draining, which would apparently
still require the related RDS configuration that was referenced and in use
by the draining state listener to be presented, or otherwise envoy
will keep trying to resolve its RDS dependencies.

This commit ensures that for a given RDS request, we always give back the
referenced RDS resources, filled with empty/stub-like RouteConfigs when
actual logic did not compute them (as they should no longer be necessary
in the new version of LDS).

Fixes #3258

- Control Plane          [x]

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
No

### CODE REVIEW

File: pkg/envoy/rds/response.go [hunk 1]
```diff
@@ -56,5 +58,40 @@ func NewResponse(cataloger catalog.MeshCataloger, proxy *envoy.Proxy, _ *xds_dis
 		rdsResources = append(rdsResources, ingressRouteConfig)
 	}
 
+	if dr != nil {
```
Comment (shashankram) [inline pkg/envoy/rds/response.go:61]: When would this be nil?

File: pkg/envoy/rds/response.go [hunk 2]
```diff
@@ -13,7 +15,7 @@ import (
 )
 
 // NewResponse creates a new Route Discovery Response.
-func NewResponse(cataloger catalog.MeshCataloger, proxy *envoy.Proxy, _ *xds_discovery.DiscoveryRequest, cfg configurator.Configurator, _ certificate.Manager) ([]types.Resource, error) {
+func NewResponse(cataloger catalog.MeshCataloger, proxy *envoy.Proxy, dr *xds_discovery.DiscoveryRequest, cfg configurator.Configurator, _ certificate.Manager) ([]types.Resource, error) {
```
Comment (shashankram) [inline pkg/envoy/rds/response.go:18]: s/dr/discoveryReq/g

**Review by shashankram — COMMENTED** [review]
Thanks, going to test this!

File: pkg/envoy/rds/response.go (hunk 1 above)
Comment (eduser25) [reply pkg/envoy/rds/response.go:61]: Actually.... nvm. Unit tests are assuming nil, but this makes the tests wrong. I'll have to fix.

File: pkg/envoy/rds/response.go [hunk 3]
```diff
@@ -1,8 +1,10 @@
 package rds
 
 import (
+	mapset "github.com/deckarep/golang-set"
 	xds_discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
 	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
+	"github.com/envoyproxy/go-control-plane/pkg/cache/v2"
```
Comment (draychev) [inline pkg/envoy/rds/response.go:7]: Would it make sense to use `v3` instead? I see that v2 is gone from main branch.

File: pkg/envoy/rds/response.go [hunk 4]
```diff
@@ -56,5 +58,40 @@ func NewResponse(cataloger catalog.MeshCataloger, proxy *envoy.Proxy, _ *xds_dis
...
+
 	return rdsResources, nil
 }
+
+// ensureRDSRequestCompletion computes delta between requested resources and response resources.
+// If any resources requested were not responded to, this function will fill those in with empty RouteConfig stubs
+func ensureRDSRequestCompletion(dr *xds_discovery.DiscoveryRequest, rdsResources []types.Resource) []types.Resource {
+	requestMapset := mapset.NewSet()
+	for idx := range dr.ResourceNames {
+		requestMapset.Add(dr.ResourceNames[idx])
```
Comment (draychev) [inline pkg/envoy/rds/response.go:76]: This should work too, right? 
```suggestion
	for _, resourceName := range dr.ResourceNames {
		requestMapset.Add(resourceName)
```

File: pkg/envoy/rds/response.go [hunk 5]
```diff
@@ -56,5 +58,40 @@ func NewResponse(cataloger catalog.MeshCataloger, proxy *envoy.Proxy, _ *xds_dis
...
+	}
+
+	responseMapset := mapset.NewSet()
+	for _, res := range rdsResources {
+		responseMapset.Add(cache.GetResourceName(res))
+	}
+
+	// If there were any requested elements we didn't reply to, create empty/dummy RDS resources
```
Comment (draychev) [inline pkg/envoy/rds/response.go:84]:
```suggestion
	// If there were any requested elements we didn't reply to, create empty RDS resources
```

File: pkg/envoy/rds/response.go [hunk 6]
```diff
@@ -56,5 +58,40 @@ func NewResponse(cataloger catalog.MeshCataloger, proxy *envoy.Proxy, _ *xds_dis
...
+	// for those now
+	requestDifference := requestMapset.Difference(responseMapset)
+	requestDifference.Each(func(reqDif interface{}) bool {
+		unfulfilledRequestedResource := reqDif.(string)
+		rdsResources = append(rdsResources, route.NewRouteConfigurationStub(unfulfilledRequestedResource))
+
+		// Do not stop iterating until last elem
+		return false
+	})
```
Comment (draychev) [inline pkg/envoy/rds/response.go:93]: I love the functional approach `mapset` provides with `Each()`.  But I wonder if we can cut 3 lines by using `.Iterator()` - we won't have to `return false`.
```suggestion
	for reqDif := requestMapset.Difference(responseMapset).Iterator() {
		unfulfilledRequestedResource := reqDif.(string)
		rdsResources = append(rdsResources, route.NewRouteConfigurationStub(unfulfilledRequestedResource))
	}
```

File: pkg/envoy/rds/response.go (hunk 3 above)
Comment (draychev) [inline pkg/envoy/rds/response.go:7]: Would it make sense to alias this with the name of the package - `cache` or even `cache_v2` or `cache_v3`?

**Review by draychev — APPROVED** [review]
Thanks for tackling this! LGTM so far.
Look forward to @shashankram's testing

File: pkg/envoy/rds/response.go (hunk 2 above)
Comment (eduser25) [reply pkg/envoy/rds/response.go:18]: :ok:

File: pkg/envoy/rds/response.go (hunk 1 above)
Comment (eduser25) [reply pkg/envoy/rds/response.go:61]: Working on this.... will take a while to fix unit tests

File: pkg/envoy/rds/response.go (hunk 3 above)
Comment (eduser25) [reply pkg/envoy/rds/response.go:7]: I should have already fixed this. this is a miss-push.

File: pkg/envoy/rds/response.go (hunk 4 above)
Comment (eduser25) [reply pkg/envoy/rds/response.go:76]: :ok_hand:

File: pkg/envoy/rds/response.go (hunk 6 above)
Comment (eduser25) [reply pkg/envoy/rds/response.go:93]: :ok_man:

File: pkg/envoy/rds/response.go (hunk 3 above)
Comment (eduser25) [reply pkg/envoy/rds/response.go:7]: misstype from VScode autocompletion

File: pkg/envoy/rds/response.go (hunk 5 above)
Comment (eduser25) [reply pkg/envoy/rds/response.go:84]: :+1:

**Review by shashankram — APPROVED** [review]
Tested this change and it works as expected. I no longer see unfulfilled requests for RDS resources, which was resulting in a continuous request-response XDS loop.

**Pull request merged**